source/extensions/filters/http/api_key_auth/api_key_auth.cc

```diff
@@ -0,0 +1,172 @@
+  } else {
+    source_ = std::string(cookie);
+  }
+}
+
+KeySources::KeySources(const KeySourcesProto& proto_config) {
+  key_sources_.reserve(proto_config.entries_size());
+  for (const auto& source : proto_config.entries()) {
+    key_sources_.emplace_back(source.header(), source.query(), source.cookie());
+  }
+}
+
+absl::string_view KeySources::Source::getKey(const Http::RequestHeaderMap& headers,
+                                             std::string& buffer) const {
+  if (absl::holds_alternative<Http::LowerCaseString>(source_)) {
+    if (const auto header = headers.get(absl::get<Http::LowerCaseString>(source_));
+        !header.empty()) {
+      absl::string_view header_view = header[0]->value().getStringView();
+      if (absl::StartsWith(header_view, "Bearer ")) {
+        header_view = header_view.substr(7);
+      }
+      std::cout << "header_view: " << header_view << std::endl;
+      return header_view;
+    }
+  } else if (query_source_) {
+    auto params =
+        Http::Utility::QueryParamsMulti::parseAndDecodeQueryString(headers.getPathValue());
+    if (auto iter = params.data().find(absl::get<std::string>(source_));
+        iter != params.data().end()) {
+      if (!iter->second.empty()) {
+        buffer = std::move(iter->second[0]);
+        return buffer;
+      }
+    }
+  } else {
+    buffer = Http::Utility::parseCookieValue(headers, absl::get<std::string>(source_));
+    return buffer;
+  }
+
+  return {};
+}
+
+absl::string_view KeySources::getKey(const Http::RequestHeaderMap& headers,
+                                     std::string& buffer) const {
+  for (const auto& source : key_sources_) {
+    if (auto key = source.getKey(headers, buffer); !key.empty()) {
+      return key;
+    }
+  }
+  return {};
+}
+
+FilterConfig::FilterConfig(const ApiKeyAuthProto& proto_config, Stats::Scope& scope,
+                           const std::string& stats_prefix)
+    : default_config_(proto_config), stats_(generateStats(scope, stats_prefix + "api_key_auth.")) {}
+
+ApiKeyAuthFilter::ApiKeyAuthFilter(FilterConfigSharedPtr config) : config_(std::move(config)) {}
+
+Http::FilterHeadersStatus ApiKeyAuthFilter::decodeHeaders(Http::RequestHeaderMap& headers, bool) {
+  const RouteConfig* override_config =
+      Http::Utility::resolveMostSpecificPerFilterConfig<RouteConfig>(decoder_callbacks_);
+
+  OptRef<const Credentials> credentials = config_->credentials();
+  OptRef<const KeySources> key_sources = config_->keySources();
+
+  // If there is an override config, then try to override the API key map and key source.
+  if (override_config != nullptr) {
+    OptRef<const Credentials> override_credentials = override_config->credentials();
+    if (override_credentials.has_value()) {
+      credentials = override_credentials;
+    }
+
+    OptRef<const KeySources> override_key_sources = override_config->keySources();
+    if (override_key_sources.has_value()) {
+      key_sources = override_key_sources;
+    }
```

Comment:
First, I want to explain why we want selective overriding.

The authentication filter is a little different with other normal filter. A well designed gateway may provide some abstractions manage the credentials. And users manage the abstraction (by credentials management module) rather then to operate the filter configuration directly. (This is the difference between the authentication filters and other filters)

If per route credentials is used, then when the credentials are updated (by the credentials management module), the control plane may need to iterate all routes to update their configuration based on the credentials update.

So, more common way is using a global filter level credentials list and per route level key sources (if necessary) for most gateways, this will simplify the control plane compexity.

But we also need to consider that there are some scenarios that require per route level crendentials, like for multiple tanants gateway.

So, finally I choose the selective overriding to do that. By this way, the users needn't to configure per route credentials list if they only want per route level key sources.

--- 

And if the selective overriding is used, I think then we shouldn't add new config knob of a primitive type. The wrapped value or message should be used here.